traffic_ops/traffic_ops_golang/server/servers_server_capability.go

```diff
@@ -0,0 +1,167 @@
+)
+
+// we need a type alias to define functions on
+type TOServerServerCapability struct {
+	api.APIInfoImpl `json:"-"`
+	tc.ServerServerCapability
+}
+
+func (ssc *TOServerServerCapability) SetLastUpdated(t tc.TimeNoMod) { ssc.LastUpdated = &t }
+func (ssc *TOServerServerCapability) NewReadObj() interface{} {
+	return &tc.ServerServerCapability{}
+}
+func (ssc *TOServerServerCapability) SelectQuery() string { return scSelectQuery() }
+func (ssc *TOServerServerCapability) ParamColumns() map[string]dbhelpers.WhereColumnInfo {
+	return map[string]dbhelpers.WhereColumnInfo{
+		ServerCapabilityQueryParam: dbhelpers.WhereColumnInfo{"sc.server_capability", nil},
+		ServerQueryParam:           dbhelpers.WhereColumnInfo{"s.id", api.IsInt},
+		ServerHostNameQueryParam:   dbhelpers.WhereColumnInfo{"s.host_name", nil},
+	}
+
+}
+func (ssc *TOServerServerCapability) DeleteQuery() string { return scDeleteQuery() }
+
+func (ssc TOServerServerCapability) GetKeyFieldsInfo() []api.KeyFieldInfo {
+	return []api.KeyFieldInfo{
+		{ServerQueryParam, api.GetIntKey},
+		{ServerCapabilityQueryParam, api.GetStringKey},
+	}
+}
+
+// Need to satisfy Identifier interface but is a no-op as path does not have Update
+func (ssc TOServerServerCapability) GetKeys() (map[string]interface{}, bool) {
+	if ssc.ServerID == nil {
+		return map[string]interface{}{ServerQueryParam: 0}, false
+	}
+	if ssc.ServerCapability == nil {
+		return map[string]interface{}{ServerCapabilityQueryParam: 0}, false
+	}
+	return map[string]interface{}{
+		ServerQueryParam:           *ssc.ServerID,
+		ServerCapabilityQueryParam: *ssc.ServerCapability,
+	}, true
+}
+
+func (ssc *TOServerServerCapability) SetKeys(keys map[string]interface{}) {
+	sID, _ := keys[ServerQueryParam].(int)
+	ssc.ServerID = &sID
+
+	sc, _ := keys[ServerCapabilityQueryParam].(string)
+	ssc.ServerCapability = &sc
+}
+
+func (ssc *TOServerServerCapability) GetAuditName() string {
+	if ssc.ServerCapability != nil {
+		return *ssc.ServerCapability
+	}
+	return "unknown"
+}
+
+func (ssc *TOServerServerCapability) GetType() string {
+	return "server server_capability"
+}
+
+// Validate fulfills the api.Validator interface
+func (ssc TOServerServerCapability) Validate() error {
+	errs := validation.Errors{
+		ServerQueryParam:           validation.Validate(ssc.ServerID, validation.Required),
+		ServerCapabilityQueryParam: validation.Validate(ssc.ServerCapability, validation.Required),
+	}
+
+	return util.JoinErrs(tovalidate.ToErrors(errs))
+}
+
+func (ssc *TOServerServerCapability) Update() (error, error, int) {
+	return nil, nil, http.StatusNotImplemented
+}
+
+func (ssc *TOServerServerCapability) Read() ([]interface{}, error, error, int) {
+	return api.GenericRead(ssc)
+}
+
+func (ssc *TOServerServerCapability) Delete() (error, error, int) {
+	return api.GenericDelete(ssc)
+}
+
+func (ssc *TOServerServerCapability) Create() (error, error, int) {
+	resultRows, err := ssc.APIInfo().Tx.NamedQuery(scInsertQuery(), ssc)
```

Comment:
We might want to do this in a follow-up PR, but I'm just going to add the comment here for awareness. There was discussion around what should happen if a server is assigned the same server_capability multiple times, and we think it should just ignore/accept the new assignment without returning an error from the API. We might just be able to check the result of `api.ParseDBError` for uniqueness constraint errors.

Reply:
I can do that! I am little worried about consistency across our endpoints. Do we support that idempotent behavior anywhere else? I am guessing not since the ParseDBError is a util method used across our handlers

Reply:
Yeah I think maybe we should hold off and possibly do it in a follow-up PR. Seems doable, but the other "many-to-many" API we have is "deliveryservice-servers" (assigning DSes to servers). That one actually has two endpoints that basically do the same thing. One _replaces_ the association by deleting the association first, and the other seems to return an error if the association is already there.